cs244b_project/src/lib.rs

```diff
@@ -583,43 +600,26 @@ impl StreamletInstance {
         }
     }
 
-    /* Verifies message signatures against all known public keys, returning the number of valid signatures
-    @param message: the message instance with signatures to be validated */
-    fn verify_message(&self, message: &Message) -> usize {
-        let mut num_valid_signatures = 0;
-        let signatures = message.clone().get_signatures(); // Check all signatures
-
-        // Check all sigatures on the message; prevent double-signing
-        let mut seen = std::collections::HashSet::new();
-        for signature in signatures.iter() {
-            // Check against all known pk's
-            for name in self.public_keys.keys() {
-                let pk = self.public_keys[name];
-                if !seen.contains(name) && self.verify_signature(message, signature, &pk) {
-                    num_valid_signatures += 1;
-                    seen.insert(name);
-                    break;
-                }
-            }
-        }
-        debug!(
-            "Attempted validation on message {}, found {} valid signatures",
-            message.nonce, num_valid_signatures
-        );
-        return num_valid_signatures;
-    }
-
-    /* Determines if the block associated with a message is notarized.
-    @param epoch: epoch number */
-    pub fn is_notarized(&self, message: &Message) -> bool {
-        // Partially synchronous model: 
-        // - Finalize after three blocks
-        // - >= 2N/3 signatures for notarization 
+    /* Determines if a given block is notarized. 
+        NOTE: assumes that all signatures locally cached are valid for the block 
+        with a hash matching the given `block.hash`. 
+        @param block: block received in a vote/proposal */
+    pub fn is_notarized(&self, block: &Block, message: &Message) -> bool {
+        // Partially synchronous model: >= 2N/3 valid signatures for notarization 
         // Note: expected peer count = excluding self; add one to get N
-        return self.verify_message(message)
```

Comment:
Instead of just checking the MESSAGE, check the total number of signatures received on this block in this epoch.